### ISSUE TITLE
recurrent state

### ISSUE THREAD

**Issue by wj210** [lucidrains/block-recurrent-transformer-pytorch#3]
Hi, i noticed that recurrent state keys and values are derived from self.init_state rather than previous recurrent state. In this case, the keys and value embedding layers, along with the projection and MLP layers for horizontal attention will not be trained? I am not sure if this is the intended behavior as I could not find any details regarding this in the paper.

**Comment by lucidrains** [comment]
@wj210 only if you don't pass in the previous state, it will use the `init_state`, much like the initial hiddens of an RNN https://github.com/lucidrains/block-recurrent-transformer-pytorch/blob/main/block_recurrent_transformer_pytorch/block_recurrent_transformer_pytorch.py#L518

**Comment by lucidrains** [comment]
@wj210 do let me know if i misunderstood something!

**Comment by wj210** [comment]
I tried checking the recurrent layers to see if it is trained, for example the projection/mlp layers and it is not. After changing the key and values to orig_state, then it trains. The code shows that the keys and values are always using self.init_state, which is hard coded in, instead of referring to orig_states.

**Comment by wj210** [comment]
`state_q, state_k, state_v = (self.state_to_q(self.init_state), *self.state_to_kv(self.init_state).chunk(2, dim = -1)) `

**Comment by lucidrains** [comment]
@wj210 :man_facepalming: my mistake! thank you!
